File: CheckForBST/CheckForBST.py
# https://www.geeksforgeeks.org/problems/check-for-bst/1

#User function Template for python3


class Solution:
    
    #Function to check whether a Binary Tree is BST or not.
    def isBST(self, root):
        
        def depthFirstSearch_forRightTree(primaryRoot, currentRoot):
            if currentRoot is None:
                return True
            if currentRoot.data <= primaryRoot.data:
                return False
            if currentRoot.left is not None:
                if currentRoot.left.data >= currentRoot.data:
                    return False
            if currentRoot.right is not None:
                if currentRoot.right.data <= currentRoot.data:
                    return False
            leftTreeCall=depthFirstSearch_forRightTree(primaryRoot, currentRoot.left)
            rightTreeCall=depthFirstSearch_forRightTree(primaryRoot, currentRoot.right)
            return leftTreeCall and rightTreeCall
        
        def depthFirstSearch_forLeftTree(primaryRoot, currentRoot):
            if currentRoot is None:
                return True
            if currentRoot.data >= primaryRoot.data:
                return False
            if currentRoot.left is not None:
                if currentRoot.left.data >= currentRoot.data:
                    return False
            if currentRoot.right is not None:
                if currentRoot.right.data <= currentRoot.data:
                    return False
            leftTreeCall=depthFirstSearch_forLeftTree(primaryRoot, currentRoot.left)
            rightTreeCall=depthFirstSearch_forLeftTree(primaryRoot, currentRoot.right)
            return leftTreeCall and rightTreeCall
        
        def isBST_postOrderTraversal(root):
            leftTreeCall=depthFirstSearch_forLeftTree(root, root.left)
            rightTreeCall=depthFirstSearch_forRightTree(root, root.right)
            return leftTreeCall and rightTreeCall
        
        return isBST_postOrderTraversal(root);




#{ 
 # Driver Code Starts
#Initial Template for Python 3
from collections import deque


# Tree Node
class Node:

    def __init__(self, val):
        self.right = None
        self.data = val
        self.left = None


# Function to Build Tree
def buildTree(s):
    #Corner Case
    if (len(s) == 0 or s[0] == "N"):
        return None

    # Creating list of strings from input
    # string after spliting by space
    ip = list(map(str, s.split()))

    # Create the root of the tree
    root = Node(int(ip[0]))
    size = 0
    q = deque()

    # Push the root to the queue
    q.append(root)
    size = size + 1

    # Starting from the second element
    i = 1
    while (size > 0 and i < len(ip)):
        # Get and remove the front of the queue
        currNode = q[0]
        q.popleft()
        size = size - 1

        # Get the current node's value from the string
        currVal = ip[i]

        # If the left child is not null
        if (currVal != "N"):

            # Create the left child for the current node
            currNode.left = Node(int(currVal))

            # Push it to the queue
            q.append(currNode.left)
            size = size + 1
        # For the right child
        i = i + 1
        if (i >= len(ip)):
            break
        currVal = ip[i]

        # If the right child is not null
        if (currVal != "N"):

            # Create the right child for the current node
            currNode.right = Node(int(currVal))

            # Push it to the queue
            q.append(currNode.right)
            size = size + 1
        i = i + 1
    return root


if __name__ == "__main__":
    t = int(input())
    for _ in range(0, t):
        s = input()
        root = buildTree(s)
        if Solution().isBST(root):
            print("true")
        else:
            print("false")

# } Driver Code Ends

File: CheckForBST/CheckForBST.cpp
/*
https://www.geeksforgeeks.org/problems/check-for-bst/1
*/

//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;
#define MAX_HEIGHT 100000

// Tree Node
struct Node {
    int data;
    Node* left;
    Node* right;

    Node(int val) {
        data = val;
        left = right = NULL;
    }
};


// } Driver Code Ends
class Solution {
  public:
  
    bool dfs2(Node *prroot, Node *root){
        if(root==nullptr) return true;
        if(root->data<=prroot->data) return false;
        if(root->left!=nullptr){
            if(root->left->data>root->data) return false;
        }
        if(root->right!=nullptr){
            if(root->right->data<root->data) return false;
        }
        bool lftTreeCall=dfs2(prroot, root->left);
        bool rightTreeCall=dfs2(prroot, root->right);
        return lftTreeCall&&rightTreeCall;
    }
  
    bool dfs1(Node *prroot, Node *root){
        if(root==nullptr) return true;
        if(root->data>=prroot->data) return false;
        if(root->left!=nullptr){
            if(root->left->data>root->data) return false;
        }
        if(root->right!=nullptr){
            if(root->right->data<root->data) return false;
        }
        bool lftTreeCall=dfs1(prroot, root->left);
        bool rightTreeCall=dfs1(prroot, root->right);
        return lftTreeCall&&rightTreeCall;
    }
  
    bool isBST_postOrderTraversal(Node *root){
        bool lftcl=dfs1(root, root->left);
        bool rghtcl=dfs2(root, root->right);
        return lftcl&&rghtcl;
    }
  
    // Function to check whether a Binary Tree is BST or not.
    bool isBST(Node* root) {
        return isBST_postOrderTraversal(root);
    }
};


//{ Driver Code Starts.

// Function to Build Tree
Node* buildTree(string str) {
    // Corner Case
    if (str.length() == 0 || str[0] == 'N')
        return NULL;

    // Creating vector of strings from input
    // string after spliting by space
    vector<string> ip;

    istringstream iss(str);
    for (string str; iss >> str;)
        ip.push_back(str);

    // Create the root of the tree
    Node* root = new Node(stoi(ip[0]));

    // Push the root to the queue
    queue<Node*> queue;
    queue.push(root);

    // Starting from the second element
    int i = 1;
    while (!queue.empty() && i < ip.size()) {

        // Get and remove the front of the queue
        Node* currNode = queue.front();
        queue.pop();

        // Get the current node's value from the string
        string currVal = ip[i];

        // If the left child is not null
        if (currVal != "N") {

            // Create the left child for the current node
            currNode->left = new Node(stoi(currVal));

            // Push it to the queue
            queue.push(currNode->left);
        }

        // For the right child
        i++;
        if (i >= ip.size())
            break;
        currVal = ip[i];

        // If the right child is not null
        if (currVal != "N") {

            // Create the right child for the current node
            currNode->right = new Node(stoi(currVal));

            // Push it to the queue
            queue.push(currNode->right);
        }
        i++;
    }

    return root;
}

void inorder(Node* root, vector<int>& v) {
    if (root == NULL)
        return;

    inorder(root->left, v);
    v.push_back(root->data);
    inorder(root->right, v);
}

int main() {

    int t;
    string tc;
    getline(cin, tc);
    t = stoi(tc);
    while (t--) {
        string s;
        getline(cin, s);
        // if (s[0] == '9') {
        //     cout << "false\n";
        //     return 0;
        // }
        Node* root = buildTree(s);
        Solution ob;
        if (ob.isBST(root))
            cout << "true\n";

        else
            cout << "false\n";
    }
    return 0;
}

// } Driver Code Ends